hyperkit: Add --networking option

MagnusS opened the pull request:
Adds a new bool option to disable networking with `linuxkit run hyperkit`.
Networking is enabled by default.

Fixes #1912 

![web_small-animal1](https://cloud.githubusercontent.com/assets/1076486/26547655/6562a0cc-4470-11e7-9dcd-e9586c4a7b00.jpg)

rneugeba commented on src/cmd/linuxkit/run_hyperkit.go:
```diff
@@ -49,6 +50,13 @@ func runHyperKit(args []string) {
+	if *vpnKitEthernetSocket == "" && *networking {
+		log.Fatalln("VPNKit socket path can't be empty when networking is enabled.")
+	}
+
```
could this be grouped closer to the rest of the networking code below?

justincormack commented:
Booleans that default to true are a bit odd in cli flags, `--disable-networking` is more idiomatic. But actually having `--networking=vpnkit` might be a better default and support `--networking=none` so could add other network options later.

rneugeba commented:
I would also prefer `--networking=`. Could we maybe also then simplify the other VPNKit options. Default to re-use Docker for Mac, but use an option to start your own VPNKit process and disable networking completely?

MagnusS commented:
@rneugeba Hm, do you mean having e.g "--networking=docker-for-mac", and that a different "vpnkit" mode would start vpnkit? We also need an option to specify a custom socket (or an existing state dir with the socket in it), so we can build networks with the new vpnkit instance

justincormack commented:
I think I would have
- `--networking=docker-for-mac` (default)
- `--networking=vpnkit` which has a socket option, maybe defaults to create, then either do `--networking=vpnkit,/path/to/sock` or `--vpnkit-socket=path`
- `--networking-none`

rneugeba commented:
yes, something along those lines. I think we want a primary networking option to select the mode (re-use DfM (default), start own VPNKit, none). The `vpnkit-socket` option is really secondary and only makes sense when we start our own VPNKit *and* want to have multiple HyperKits sharing it, which is probably a bit niche.

Ah, just saw @justincormack's update. I think we propose the same thing. I prefer the `--networking=vpnkit,/path/to/sock` variation as it is one less command line options

MagnusS commented:
I like that too - much cleaner. But I guess we actually have two vpnkit modes, one for starting your own instance that uses a hardcoded socket path (`[state]/vpnkit_eth.sock`), and one for joining an existing one which currently takes a path to a socket. We could simplify that, e.g.:

`--networking=vpnkit`
`--networking=vpnkit-join,state-dir-of-first-node`

...and remove other vpnkit options

rneugeba commented:
or could you check if the socket exists and then assume join?

justincormack commented:
We already have a `state-dir` option, so I think omitting the path can imply that you start your own instance in state-dir, while specifying it means join an existing one at the specified path.

rneugeba commented:
`state-dir` also contains per VM state so it's different and can't be shared between VMs

justincormack commented:
No, I meant if you want to set where the path is for the master you have state-dir, so no need to specify the socket location explicitly, while for the case of using existing socket you can specify the path.

rneugeba commented:
sorry, misread. the above makes sense

MagnusS commented:
@rneugeba @justincormack thanks for comments, I pushed a new version that supports the following options:
`--networking=docker-for-mac` -- default, connects to DfM
`--networking=vpnkit` -- launch new vpnkit
`--networking=vpnkit,/.../other-vm-state-dir` -- connect to existing vpnkit
`--networking=none`

Also removed the `--vpnkit-socket` option and moved the networking code to the same place

MagnusS commented:
Ah, small bug with setting up the ports from the `vsock-ports` option too late so they overwrote the vpnkit forwarding port - fixed in latest update

justincormack commented:
The option should point at the socket file not the state dir I think.

On 29 May 2017 15:33, "Magnus Skjegstad" <notifications@github.com> wrote:

> Ah, small bug with setting up the ports from the vsock-ports option too
> late so they overwrote the vpnkit forwarding port - fixed in latest update
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/linuxkit/linuxkit/pull/1926#issuecomment-304675958>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAdcPI1JSG5hYzuQbk-L6H1v74_HhJaMks5r-tcygaJpZM4NpL9O>
> .
>

MagnusS commented:
@justincormack ah ok, then I misread too :-) I can change it to point to socket. You can't override it now so it will always have the same name relative to the state dir though (unless you launch vpnkit in some other way)

justincormack commented:
@MagnusS yes thats fine, but you might well start vpnkit yourself in which case there is no concept of state dir, so its less confusing. We just need to document the socket name when you use `linuxkit` to create the socket (and make sure its output in verbose mode).

rneugeba commented on src/cmd/linuxkit/run_hyperkit.go:
```diff
@@ -147,20 +164,18 @@ func runHyperKit(args []string) {
+	networkingDockerForMac        = "docker-for-mac"
+	networkingVPNKit              = "vpnkit"
+)
+
```
I don't know what's the idiomatic go way is but I find it odd to only define these constants after prominent use further up. I would define them after the import section at the beginning of the file...

MagnusS replied:
Ok - moved to the top
